### DISCUSSION TITLE
Toyota Mirai Sensors

### DISCUSSION THREAD

**Discussion by tchicken** [DurgNomis-drol/ha_toyota#66] [Ideas]
I know this isn't the most common of cars, but I've just taken delivery of a new Toyota Mirai hydrogen fuel cell car and have connected it into Home Assistant using this integration. 

There are a few additional sensors that I would love to see working, namely range remaining and level %, and also the HVAC information.

Using the code in the RAV4 thread I've gathered the following:

```
DEBUG:mytoyota:Raw status data: {'VehicleInfo': {'AcquisitionDatetime': '2021-09-19T14:03:47Z', 'RemoteHvacInfo': {'InsideTemperature': 26, 'RemoteHvacMode': 0, 'RemoteHvacProhibitionSignal': 1, 'Temperaturelevel': 33, 'SettingTemperature': 20.0, 'BlowerStatus': 0, 'FrontDefoggerStatus': 0, 'RearDefoggerStatus': 0, 'FrontSeatHeaterStatus': 1, 'RearSeatHeaterStatus': 0, 'SteeringHeaterStatus': 0, 'FunctionInfoFixedStatus': 1, 'LatestAcStartTime': '2021-09-17T20:04:17Z', 'TemperatureDisplayFlag': 1}, 'ChargeInfo': {'GasolineTravelableDistanceUnit': 2, 'GasolineTravelableDistance': 183, 'EvDistanceInKm': 0.0, 'EvTravelableDistance': 0.0}}, 'ReturnCode': '000000'}
DEBUG:httpx._client:HTTP Request: GET https://myt-agg.toyota-europe.com/cma/api/users/ecd0a9cf-21fa-47c9-ab0f-5bf0f1eb69b1/vehicles/XXXXXXXXXXXXXX/vehicleStatus "HTTP/1.1 200 OK"
{'event': {'lat': 'XXXXXXXXX', 'lon': 'XXXXXXXXX', 'timestamp': '1632060096000'}, 'tripStatus': '0', 'energy': [{'timestamp': '2021-09-19T14:02:37Z', 'type': 'HYDROGEN', 'level': 58.8, 'remainingRange': 295}], 'protectionState': {'overallStatus': 'OK', 'timestamp': '2021-09-19T14:06:55Z', 'doors': {'warning': False, 'driverSeatDoor': {'warning': False, 'closed': True, 'locked': True}, 'passengerSeatDoor': {'warning': False, 'closed': True, 'locked': True}, 'rearRightSeatDoor': {'warning': False, 'closed': True, 'locked': True}, 'rearLeftSeatDoor': {'warning': False, 'closed': True, 'locked': True}}, 'hood': {'closed': True, 'warning': False}, 'lamps': {'warning': False, 'headLamp': {'warning': False, 'off': True}, 'tailLamp': {'warning': False, 'off': True}, 'hazardLamp': {'warning': False, 'off': True}}, 'windows': {'warning': False, 'driverSeatWindow': {'warning': False, 'state': 'close'}, 'passengerSeatWindow': {'warning': False, 'state': 'close'}, 'rearRightSeatWindow': {'warning': False, 'state': 'close'}, 'rearLeftSeatWindow': {'warning': False, 'state': 'close'}}, 'key': {'warning': False, 'inCar': False}, 'lock': {'lockState': 'locked', 'source': 'key', 'failedUnlockPreconditions': ['alarmLocked']}}}
```

The part I'm most interested in is this:
`'energy': [{'timestamp': '2021-09-19T14:02:37Z', 'type': 'HYDROGEN', 'level': 58.8, 'remainingRange': 295}]
`

level is % and range is km.

I may be able to contribute in the future, but I don't have the time at the moment. If you're able to add the information that would be great. I'm assuming that some of the information will also be used in future Toyota cars so will probably be of use for others.

Thanks,

**Comment by DurgNomis-drol** [comment]
Created an Issue for it so we can track it there. We will start with energy sensors #67 and look into the HVAC sensors afterwards #68. 

It needs to be implemented in `mytoyota` first before we can work on implementing it here. 

I actually hoped for someone to turn up here requesting that information as I stumbled upon it when looking through the Android App's source code. Thanks for supplying a lot of information upfront, it helps a lot and makes it a lot easier.

And congrats with your new car!! It is cool that it uses a hydrogen fuel cell :sunglasses:

**Reply by tchicken** [reply]
Thanks for the really quick reply on this. If you need any more information please ask and I'll try and provide it.

**Reply by DurgNomis-drol** [reply]
Just merged the changes needed to make the energy part possible. 

The tough part is the HVAC information. Can I get you to change a setting like blower, HVAC mode, Seat heaters and so, and make a call to the endpoint between. So i can analyse me to what the different things means?

**Reply by tchicken** [reply]
I've tried playing around with settings and recording the data. Updates appear to be very random and some data doesn't match the car. 

Car off, temperature set to 19degC, heated front seats set to auto:
```
DEBUG:mytoyota:Raw status data: {'VehicleInfo': {'AcquisitionDatetime': '2021-09-23T14:02:57Z', 'RemoteHvacInfo': {'InsideTemperature': 25, 'RemoteHvacMode': 0, 'RemoteHvacProhibitionSignal': 1, 'Temperaturelevel': 32, 'SettingTemperature': 19.0, 'BlowerStatus': 0, 'FrontDefoggerStatus': 0, 'RearDefoggerStatus': 0, 'FrontSeatHeaterStatus': 1, 'RearSeatHeaterStatus': 0, 'SteeringHeaterStatus': 0, 'FunctionInfoFixedStatus': 1, 'LatestAcStartTime': '2021-09-23T11:40:17Z', 'TemperatureDisplayFlag': 1}, 'ChargeInfo': {'EvTravelableDistance': 0.0, 'GasolineTravelableDistance': 259, 'GasolineTravelableDistanceUnit': 2, 'EvDistanceInKm': 0.0}}, 'ReturnCode': '000000'}
```

From the app, remotely set:

Temperature to 25degC
Front seats on
Front heated screen on 
Rear heated screen on

```
DEBUG:mytoyota:Raw status data: {'VehicleInfo': {'ChargeInfo': {'EvDistanceInKm': 0.0, 'GasolineTravelableDistanceUnit': 2, 'GasolineTravelableDistance': 259, 'EvTravelableDistance': 0.0}, 'AcquisitionDatetime': '2021-09-23T11:36:40Z', 'RemoteHvacInfo': {'InsideTemperature': 20, 'RemoteHvacMode': 1, 'RemoteHvacProhibitionSignal': 0, 'Temperaturelevel': 27, 'SettingTemperature': 25.0, 'BlowerStatus': 3, 'FrontDefoggerStatus': 1, 'RearDefoggerStatus': 1, 'FrontSeatHeaterStatus': 1, 'RearSeatHeaterStatus': 0, 'SteeringHeaterStatus': 0, 'FunctionInfoFixedStatus': 1, 'LatestAcStartTime': '2021-09-23T11:36:09Z', 'TemperatureDisplayFlag': 1, 'RemainingMinutes': 19}}, 'ReturnCode': '000000'}
```

Car off, driver side set to 20degC, passenger side set to 19degC. Heated seats off:
```
DEBUG:mytoyota:Raw status data: {'VehicleInfo': {'AcquisitionDatetime': '2021-09-23T14:40:42Z', 'RemoteHvacInfo': {'TemperatureDisplayFlag': 1, 'RemoteHvacMode': 0, 'RemoteHvacProhibitionSignal': 1, 'Temperaturelevel': 30, 'SettingTemperature': 20.0, 'BlowerStatus': 0, 'FrontDefoggerStatus': 0, 'RearDefoggerStatus': 0, 'FrontSeatHeaterStatus': 1, 'RearSeatHeaterStatus': 0, 'SteeringHeaterStatus': 0, 'FunctionInfoFixedStatus': 1, 'LatestAcStartTime': '2021-09-23T11:40:17Z', 'InsideTemperature': 23}, 'ChargeInfo': {'GasolineTravelableDistance': 258, 'EvDistanceInKm': 0.0, 'GasolineTravelableDistanceUnit': 2, 'EvTravelableDistance': 0.0}}, 'ReturnCode': '000000'}
```

**Reply by DurgNomis-drol** [reply]
Thanks!!

Yeah some of the data does not seems to make much sense (As you have pointed out). 

I will limit what is exposed right now, until someone looks into implementing setting the different settings from this package (I won't be me doing it as i don't have a car that have these features, and it keeps getting more and more difficult to implement features that i can't test on my own car).

I am aiming to have something you can test in a couple of days 🚀

**Reply by DurgNomis-drol** [reply]
@tchicken Can I get you to update `mytoyota` to the latest one (0.7.0) and the run the code snippet on the README and then pass the result here. Please remember to censor your data. This will help me make sure i am not missing something 😃 Thanks in advance

**Comment by DurgNomis-drol** [comment]
@tchicken Have you seen my comment?

**Comment by tchicken** [comment]
Sorry, yes I did. I started working on the data but had a couple of issues with the update. Will try and get something to you shortly.

**Reply by tchicken** [reply]
This is what I get when I run the README code:

```
['europe']
Logging in...
Retrieving cars...
Traceback (most recent call last):
  File "mirai.py", line 86, in <module>
    loop.run_until_complete(get_information())
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "mirai.py", line 26, in get_information
    vehicle = await client.get_vehicle_status(car)
  File "/home/tchicken/toyota/toyota_venv/lib/python3.8/site-packages/mytoyota/client.py", line 129, in get_vehicle_status
    car = Vehicle(
  File "/home/tchicken/toyota/toyota_venv/lib/python3.8/site-packages/mytoyota/vehicle.py", line 107, in __init__
    self.sensors = Sensors(status.get("protectionState", {}))
  File "/home/tchicken/toyota/toyota_venv/lib/python3.8/site-packages/mytoyota/status.py", line 49, in __init__
    self.doors = Doors(status[DOORS]) if DOORS in status else None
  File "/home/tchicken/toyota/toyota_venv/lib/python3.8/site-packages/mytoyota/sensors.py", line 68, in __init__
    self.trunk = Door(doors.get("backDoor", None))
  File "/home/tchicken/toyota/toyota_venv/lib/python3.8/site-packages/mytoyota/sensors.py", line 35, in __init__
    self.warning = door.get(WARNING, None)
AttributeError: 'NoneType' object has no attribute 'get'
```

**Reply by DurgNomis-drol** [reply]
Thanks, that is a bug. I haven't accounted for if the car does not have a trunk sensor. This should be a quick fix, will look into it later or tomorrow 😊 Sorry for the inconvenience

**Reply by DurgNomis-drol** [reply]
@tchicken I just released a new version, this should fix it, though i can't test it for my self. Can you please try again with the latest version (0.7.1) 😃 That would be greatly appreciated.

**Comment by tchicken** [comment]
That fixed that problem...

```
['europe']
Logging in...
Retrieving cars...
Traceback (most recent call last):
  File "mirai.py", line 86, in <module>
    loop.run_until_complete(get_information())
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "mirai.py", line 47, in get_information
    fuel = vehicle.odometer.fuel
AttributeError: 'Odometer' object has no attribute 'fuel'
```

**Reply by DurgNomis-drol** [reply]
Forgot that the example in the readme is not updated and also contains a lot of unecessary things for testing this. 

On mobile right now, but this one should do the trick

```python

import asyncio
from mytoyota.client import MyT

username = "jane@doe.com"
password = "MyPassword"
locale = "da-dk"

client = MyT(username=username, password=password, locale=locale, region="europe")


async def get_information():
    print("Logging in...")
    await client.login()

    print("Retrieving cars...")
    # Returns cars registered to your account + information about each car.
    cars = await client.get_vehicles()

    for car in cars:

        # Returns live data from car/last time you used it as an object.
        vehicle = await client.get_vehicle_status(car)
        
        print(vehicle.as_dict())


loop = asyncio.get_event_loop()
loop.run_until_complete(get_information())
loop.close()

**Comment by tchicken** [comment]
`{'id': 2609854, 'alias': 'Mirai', 'vin': 'XXXXXXX', 'details': {'brand': 'TOYOTA', 'deviceAddedOn': '2021-09-16T19:59:01.449Z', 'deviceId': 'XXXXXX', 'deviceTypeName': 'ToyotaTouch2Plus', 'engine': 'Fuel cell 2nd gen', 'entitledBy': 'CUSTOMER', 'entitledOn': '2021-09-17T10:39:31.947Z', 'exteriorColour': '219 ', 'firstRegistrationDate': '2021-09-15', 'fuel': 'HYDROGEN', 'generation': 'MM17PLUS', 'grade': 'Standard', 'horsePower': 182, 'hybrid': True, 'imageUrl': 'https://dj3z27z47basa.cloudfront.net/a0978670-5a8e-41cb-873a-552e5a1aef31', 'interiorColour': '20', 'isEntitled': True, 'isNC': False, 'licensePlate': 'XXXX XXX', 'modelCode': 'MI', 'modelDescription': 'Mirai', 'modelDocumentId': '12345', 'modelName': 'Mirai', 'navSubType': 'Low', 'navType': 'TOY', 'owner': False, 'ownerFlag': False, 'productionYear': '2021', 'source': 'NMSC', 'transmission': 'Q710', 'transmissionType': 'Q710', 'vehicleAddedOn': '2021-09-17T10:24:56.380Z'}, 'status': {'energy': {'level': 74.1, 'range': 367, 'type': 'Hydrogen', 'last_updated': '2021-10-03T12:32:59Z'}, 'hvac': {'legacy': True, 'current_temperature': 22, 'target_temperature': 19.0, 'blower_on': 0, 'front_defogger_on': 0, 'rear_defogger_on': 0, 'last_updated': '2021-10-02T10:20:11Z'}, 'odometer': {'mileage': 1090, 'unit': 'mi'}, 'parking': {'latitude': XX.XXXXXX, 'longitude': XX.XXXXXX, 'timestamp': 1633264318000}, 'vehicle': {'overallstatus': 'OK', 'lights': {'warning': False, 'front': {'warning': False, 'off': True}, 'back': {'warning': False, 'off': True}, 'hazard': {'warning': False, 'off': True}}, 'hood': {'warning': False, 'closed': True}, 'doors': {'warning': False, 'driverseat': {'warning': False, 'closed': True, 'locked': True}, 'passengerseat': {'warning': False, 'closed': True, 'locked': True}, 'rightrearseat': {'warning': False, 'closed': True, 'locked': True}, 'leftrearseat': {'warning': False, 'closed': True, 'locked': True}, 'trunk': {'warning': None, 'closed': None, 'locked': None}}, 'windows': {'warning': False, 'driverseat': {'warning': False, 'state': 'close'}, 'passengerseat': {'warning': False, 'state': 'close'}, 'rightrearseat': {'warning': False, 'state': 'close'}, 'leftrearseat': {'warning': False, 'state': 'close'}}, 'key': {'warning': False, 'in_car': False}, 'last_updated': '2021-10-03T12:38:29Z'}}, 'servicesEnabled': {'connectedServices': True}, 'statistics': {}}`

**Reply by DurgNomis-drol** [reply]
Thanks, everything looks good 🎉 

Please try out latest master (ha_toyota) as I have now added a range sensor and a EV sensor, though the last one you don't have much use of. HVAC sensors will follow shortly.

Can you please report back if the range sensors state is in KM or mi, as i can see your car reports miles from the odometer. I will add a conversion if needed then. Right now it reports in what ever the car reports.

**Reply by tchicken** [reply]
Range and tank % are coming through now - thank you.

Range is being shown as km so will need converting to miles. 

Thanks,

**Comment by DurgNomis-drol** [comment]
#73 Have been put on hold as we are adding python tests and more to the underlying module (`mytoyota`). This will make it so much easier to make sure that nothing breaks moving forward.

**Reply by DurgNomis-drol** [reply]
@tchicken 

Just merged in #73. This also include the updated `mytoyota` module that fixes range not being convert.

Can you please try installing master and report back if there is any bugs. Lets hope not, but can't promise. That would be a huge help 😃
